### PR TITLE
Ta64 4

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This PR finally enables running 64-bit TAs

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Xtest fails due to regression vs sanity, fixed in https://github.com/OP-TEE/optee_test/pull/56

**Comment by jforissier** [comment]
I have tested this PR on HiKey, I got two failures in the "extended xtest" suite:
```
root@HiKey:/ xtest 8009
[...]
* XTEST_TEE_8009 a7-2b-1f
./xml/include/xml_internal_api.h:495: Expression "op.params[0].memref.size == expectedDataSize" (8 == 4) is false
  XTEST_TEE_8009 FAILED
[...]
root@HiKey:/ xtest 8062
[...]
* XTEST_TEE_8062 a7-c0-a0
./xml/include/xml_internal_api.h:495: Expression "op.params[0].memref.size == expectedDataSize" (8 == 4) is false
  XTEST_TEE_8062 FAILED
```
All is fine with 32-bit TAs. I'll take a closer look tomorrow.

File: core/arch/arm/kernel/elf_load.c [hunk 1]
```diff
@@ -333,21 +542,52 @@ TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
...
+			res = copy_to(state, dst, state->vasize,
+				      phdr[n].p_vaddr, phdr[n].p_offset,
+				      phdr[n].p_filesz);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+		shoff = ehdr->e_shoff;
+		shsize = ehdr->e_shnum * sizeof(Elf32_Shdr);
+	} else {
+		Elf64_Ehdr *ehdr = state->ehdr;
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/elf_load.c]: the `then` and the `else` part are identical apart from the type of the pointers. Would it be possible to have a kind of template or macro to commonalize the code. If it becomes too complex / unreadable, it is not worth.

File: core/arch/arm/kernel/elf_load.c [hunk 2]
```diff
@@ -372,14 +611,34 @@ TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
...
+				}
+			}
+		} else {
+			Elf64_Ehdr *ehdr = state->ehdr;
+			Elf64_Shdr *shdr = state->shdr;
+
+			/* Process relocation */
+			for (n = 0; n < ehdr->e_shnum; n++) {
+				if (shdr[n].sh_type == SHT_RELA) {
+					res = e64_process_rela(state, n,
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/elf_load.c]: same remark as above on the macro, adding the change in the function exx_process_rela

File: core/arch/arm/kernel/thread_a64.S [hunk 3]
```diff
@@ -36,11 +36,11 @@
 #include "thread_private.h"
 
 	.macro get_thread_ctx core_local, res, tmp0, tmp1
-		ldr	\tmp0, [\core_local, \
+		ldr	w\tmp0, [\core_local, \
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/thread_a64.S]: was called with `x`, and is now called with `w`. Is it correct?

File: core/arch/arm/kernel/elf_load.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: I see your point, but putting this in a macro would make the code harder to read in my opinion.
I'll make an attempt with an access function to read out content of a Phdr based on `is_32bit`.

File: core/arch/arm/kernel/elf_load.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/elf_load.c]: I'll make an attempt with an access function as above.

File: core/arch/arm/kernel/thread_a64.S (hunk 3 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread_a64.S]: Yes, we're supposed to read a 32-bit value.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jenswi-linaro** [comment]
It seems checkpatch doesn't agree with my DO_ACTION macro.

**Comment by jforissier** [comment]
FYI, about the xtest failures I reported above: the bug(s) are in the test suite. I have a quick fix I need to clean before submitting a PR to `optee_test`. With the fix:

```
root@HiKey:/ xtest -l 15
[...]
48299 subtests of which 0 failed
702 test cases of which 0 failed
0 test case was skipped
TEE test application done!
root@HiKey:/ 
```

So, for both 32-bit and 64-bit TAs:
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

**Comment by pascal-brand-st-dev** [comment]
With or without my comments

    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by jenswi-linaro** [comment]
Tags applied

**Pull request merged**